Source/Core/Core/IOS/Network/KD/Mail/WC24Send.cpp

```diff
@@ -61,6 +64,118 @@ bool WC24SendList::CheckSendList() const
+}
+
+ErrorCode WC24SendList::DeleteMessage(u32 index)
+{
+  ErrorCode error = NWC24::DeleteFileFromVFF(NWC24::Mail::SEND_BOX_PATH, GetMailPath(index), m_fs);
+  if (error != WC24_OK)
+    return error;
+
+  // Fix up the header then set the entry to 0.
+  m_data.header.number_of_mail = Common::swap32(Common::swap32(m_data.header.number_of_mail) - 1);
+  m_data.header.next_entry_offset = Common::swap32(128 + (index * 128));
+  m_data.header.next_entry_id = Common::swap32(Common::swap32(m_data.header.next_entry_id) - 1);
+  m_data.header.total_size_of_messages =
+      Common::swap32(m_data.header.total_size_of_messages) - GetMailSize(index);
+
+  std::memset(&m_data.entries[index], 0, sizeof(MailListEntry));
+  return WC24_OK;
+}
+
+std::string WC24SendList::GetMailPath(u32 index) const
+{
+  return fmt::format("mb/s{:07d}.msg", GetEntryId(index));
+}
+
+u32 WC24SendList::GetNextEntryId() const
+{
+  return Common::swap32(m_data.header.next_entry_id);
+}
+
+u32 WC24SendList::GetNextEntryIndex() const
+{
+  return (Common::swap32(m_data.header.next_entry_offset) - 128) / 128;
+}
+
+std::vector<u32> WC24SendList::GetMailToSend() const
+{
+  // The list is not guaranteed to have all entries consecutively.
+  // As such we must find the populated entries for the specified number of mails.
+  const u32 mail_count = std::min(GetNumberOfMail(), 16U);
+  u32 found{};
+
+  std::vector<u32> mails{};
+  for (u32 index = 0; index < MAX_ENTRIES; index++)
+  {
+    if (found == mail_count)
+      break;
+
+    if (GetEntryId(index) != 0)
+    {
+      mails.emplace_back(index);
+      found++;
+    }
+  }
+
+  return mails;
+}
+
+ErrorCode WC24SendList::AddRegistrationMessages(const WC24FriendList& friend_list, u64 sender)
+{
+  // It is possible that the user composed a message before SendMail was called.
+  ReadSendList();
+
+  const std::vector<u64> unconfirmed_friends = friend_list.GetUnconfirmedFriends();
+  for (const u64 code : unconfirmed_friends)
+  {
+    const u32 entry_index = GetNextEntryIndex();
+    const u32 msg_id = GetNextEntryId();
+    m_data.entries[entry_index].id = Common::swap32(msg_id);
+
+    std::time_t t = std::time(nullptr);
+
+    const std::string formatted_message =
+        fmt::format(MAIL_REGISTRATION_STRING, sender, code, fmt::gmtime(t));
+    std::vector<u8> message{formatted_message.begin(), formatted_message.end()};
+    NWC24::ErrorCode reply =
+        NWC24::WriteToVFF(NWC24::Mail::SEND_BOX_PATH, GetMailPath(entry_index), m_fs, message);
+
+    if (reply != WC24_OK)
+    {
+      ERROR_LOG_FMT(IOS_WC24, "Error writing registration message to VFF");
+      return reply;
+    }
+
+    NOTICE_LOG_FMT(IOS_WC24, "Issued registration message for Wii Friend: {}", code);
+
+    // Update the header and some fields in the body
+    m_data.entries[entry_index].msg_size = Common::swap32(static_cast<u32>(message.size()));
+    m_data.header.next_entry_id = Common::swap32(entry_index + 1);
```

Comment:
This has to be bounded to `MAX_ENTRIES` somehow.

Reply:
Just realized that it should be, however the method I used for this is lazy and just assuming the next available entry is free. I should find the next available entry use that index.